### PR TITLE
Reduce allocations in `String::sprintf`

### PR DESCRIPTION
**Pull request by stuartcarnie**
This prevents an allocation of a new string for every `+= char32_t`. The example string was rendered as a label every frame in GDScript:

```gdscript
text = "%d FPS (%.2f mspf)" % [Engine.get_frames_per_second(), 1000.0 / Engine.get_frames_per_second()]
```

To validate the improvement, allocations were captured for a 5-second window. Note the number of allocations in the `sprintf` function of the following images.

**Before: 43,830 allocations**

CleanShot 2024-07-20 at 14 49 03@2x

**After `operator += (char32_t)`: 17,474 allocations**

> [!NOTE]
>
> 60% reduction in allocations.

CleanShot 2024-07-20 at 14 49 28@2x

**After `static const String`: 12,934**

> [!NOTE]
>
> A further improvement was to use some `static const String` for common padding and sign characters, which reduced the allocations a further 26%

CleanShot 2024-07-20 at 15 23 59@2x

> [!NOTE]
>
> Total reduction in allocations for the 5-second window was **70%**

### CODE REVIEW

**Comment by kus04e4ek** [comment]
Related: #93697

**Comment by kus04e4ek** [comment]
I feel like intead of `static const String`, methods like `insert_char` should be introduced, similar to #92475

**Comment by stuartcarnie** [comment]
> I feel like intead of `static const String`, methods like `insert_char` should be introduced, similar to #92475

Totally agree. Adding appending APIs would be helpful. Also, being able to keep the memory around for a String that could be used as a temporary would be beneficial too, as there are a number of temporaries created in `sprintf`

**Comment by kus04e4ek** [comment]
> Also, being able to keep the memory around for a `String` that could be used as a temporary would be beneficial too, as there are a number of temporaries created in `sprintf`

There's quite a lot of examples of `thread_local LocalVector` used for similar purposes, so I wonder about introducing a `LocalString` class, which would use `LocalVector` inside of it. Not sure of it would be worth it if there's not many uses for it though

**Comment by RandomShaper** [comment]
Just passing by. A few thoughts:
- `T const &` is inconsistent with the rest of codebase. `const T &` is preferred.
- There is string appening functionality, `StringBuilder`. I'm not sure how to use it for the late sign insertion, but there's likely a clever way to solve that.
- `String str = "("` deserves the optimization, too.


> There's quite a lot of examples of `thread_local LocalVector` used for similar purposes, so I wonder about introducing a `LocalString` class, which would use `LocalVector` inside of it. Not sure if it would be worth it if there's not many uses for it though

The `thread_local` thing is being done with `LocalVector` because it features a capacity feature, which `Vector` doesn't. In addition, `LocalVector` doesn't have a thread-safe refcount, which makes it more performant and the right option when you know you don't need such a thing. My point is that `thread_local Vector` and `thread_local String` would be possible if `Vector` featured the aforementioned capacity-reserve feature. Adding it would avoid having to emulate it in certain parts of the engine where the `Vector` is sized to what it would be the capacity and the size is managed externally.

In other words, at some point my view is that it would be ideal that `Vector` and `LocalVector` were just two specializations of the same template, with better names. So, `Vector` would be `LocalVector` plus copy-on-write
I opened a proposal time ago about this: https://github.com/godotengine/godot-proposals/issues/5144.

We could even have a copy-on-write specialization which is not thread-safe, which may be a better fit for strings. I'd say that all of this would need to be discussed carefully because there are things yelling to be improved, such as these vector types getting more telling names and finding out the most precise set of features each needs across the codebase.

**Comment by clayjohn** [comment]
Discussion of future improvements aside, I think we can go ahead and merge this once it is updated to use `const T &` formatting as requested by RandomShaper.

**Comment by stuartcarnie** [comment]
> * `T const &` is inconsistent with the rest of codebase. `const T &` is preferred.

I'll fix that.

As an aside, I wrote a blog about `const` 13 years ago, indicating a more intuitive placement, with a simple rule for reading it:

http://blog.stuartcarnie.com/post/constant-confusion/

**Comment by stuartcarnie** [comment]
@RandomShaper / @clayjohn changed to `const T &`

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected.

**Testing project:** [test_pr_94558.zip](https://github.com/user-attachments/files/16368402/test_pr_94558.zip)

## Benchmark

<details open>

<summary>PC specifications</summary>

- **CPU:** Intel Core i9-13900K
- **GPU:** NVIDIA GeForce RTX 4090
- **RAM:** 64 GB (2×32 GB DDR5-5800 C30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Linux (Fedora 40)
</details>

Using an `optimize=speed lto=full` editor build.

Time taken to format the string 1 million times from GDScript (as a standalone expression):

```gdscript
"%d FPS (%.2f mspf)" % [Engine.get_frames_per_second(), 1000.0 / Engine.get_frames_per_second()]
```

Before | After
-|-
1.009s | 0.679s

**Comment by clayjohn** [comment]
> As an aside, I wrote a blog about `const` 13 years ago, indicating a more intuitive placement, with a simple rule for reading it:

Lol, if I had to debug an issue arising from `const int const * const * i = &j; // ok` I would cry

**Review by clayjohn — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
